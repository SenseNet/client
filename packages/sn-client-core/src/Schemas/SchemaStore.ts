import { FieldSetting, GenericContent, Schema } from '@sensenet/default-content-types'

/**
 * Class that stores schema information
 */
export class SchemaStore {
  private schemas: Schema[] = []
  private byNameSchemaCache: Map<string, Schema> = new Map()

  /**
   * Updates the schema information in the store and inv
   */
  public setSchemas(newSchemas: Schema[]) {
    this.schemas = newSchemas
    this.byNameSchemaCache = new Map<string, Schema>()
  }

  /**
   * Returns the Content Type Schema for the provided Content Type;
   * @param type {string} The name of the Content Type;
   * @returns {Schemas.Schema}
   * ```ts
   * const genericContentSchema = SenseNet.Content.getSchema(Content);
   * ```
   */
  public getSchema<TType>(currentType: new (...args: any[]) => TType): Schema {
    return this.getSchemaByName(currentType.name)
  }

  private mergeFieldSettings(
    currentFieldSettings: FieldSetting[],
    parentFieldSettings: FieldSetting[],
  ): FieldSetting[] {
    const currentFieldSettingsMap = new Map<string, FieldSetting>()
    currentFieldSettings.forEach(s => currentFieldSettingsMap.set(s.Name, s))

    const parentFieldSettingsMap = new Map<string, FieldSetting>()
    parentFieldSettings.forEach(s => parentFieldSettingsMap.set(s.Name, s))

    const keys = new Set([...currentFieldSettingsMap.keys(), ...parentFieldSettingsMap.keys()])

    return Array.from(keys).map(key => {
      return {
        ...parentFieldSettingsMap.get(key),
        ...currentFieldSettingsMap.get(key),
      } as FieldSetting
    })
  }

  /**
   * Returns the Content Type Schema for the provided content type name
   * @param {string} contentTypeName The name of the content type
   */
  public getSchemaByName(contentTypeName: string) {
    if (this.byNameSchemaCache.has(contentTypeName)) {
      return Object.assign({}, this.byNameSchemaCache.get(contentTypeName)) as Schema
    }
    let schema = this.schemas.find(s => s.ContentTypeName === contentTypeName) as Schema
    if (!schema) {
      return this.getSchema(GenericContent)
    }
    schema = Object.assign({}, schema)
    const parentSchema = schema.ParentTypeName && this.getSchemaByName(schema.ParentTypeName)

    if (parentSchema) {
      schema.FieldSettings = this.mergeFieldSettings(schema.FieldSettings, parentSchema.FieldSettings)
    }
    this.byNameSchemaCache.set(contentTypeName, schema)
    return Object.assign({}, schema)
  }

  /**
   * Returns a boolean value that indicates if the specified content is an instance or descendant of a given content constructor
   * @param content The given content to check
   * @param contentType The content type constructor
   */
  public isContentFromType<T extends typeof GenericContent>(
    content: GenericContent,
    contentType: T,
  ): content is InstanceType<T> {
    if (content.Type === contentType.name) {
      return true
    }

    let currentSchema = this.getSchemaByName(content.Type)
    do {
      if (currentSchema.ContentTypeName === contentType.name) {
        return true
      }
      currentSchema = this.getSchemaByName(currentSchema.ParentTypeName || '')
    } while (currentSchema.ContentTypeName && currentSchema.ContentTypeName !== GenericContent.name)
    return contentType.name === GenericContent.name
  }

  /**
   * Returns if a given content type is a descendant of an another content type
   * @param child The child content name
   * @param parent The parent content name
   */
  public schemaIsDescendantOf(child: string, parent: string) {
    let currentSchema = this.getSchemaByName(child)
    do {
      if (currentSchema.ContentTypeName === parent) {
        return true
      }
      currentSchema = this.getSchemaByName(currentSchema.ParentTypeName || '')
    } while (currentSchema.ContentTypeName && currentSchema.ContentTypeName !== GenericContent.name)

    return parent === GenericContent.name
  }
}
