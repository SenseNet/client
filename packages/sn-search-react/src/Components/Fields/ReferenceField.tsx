import { ListItemText } from '@material-ui/core'
import CircularProgress from '@material-ui/core/CircularProgress'
import InputAdornment from '@material-ui/core/InputAdornment'
import List from '@material-ui/core/List'
import ListItem from '@material-ui/core/ListItem'
import Paper from '@material-ui/core/Paper'
import TextField, { TextFieldProps as MaterialTextFieldProps } from '@material-ui/core/TextField'
import Typography from '@material-ui/core/Typography'
import { debounce } from '@sensenet/client-utils'
import { GenericContent, ReferenceFieldSetting } from '@sensenet/default-content-types'
import { Query, QueryExpression, QueryOperators } from '@sensenet/query'
import match from 'autosuggest-highlight/match'
import parse from 'autosuggest-highlight/parse'
import React from 'react'
import Autosuggest, {
  GetSuggestionValue,
  InputProps,
  OnSuggestionSelected,
  RenderInputComponent,
  RenderSuggestion,
} from 'react-autosuggest'

/**
 * Props for the ReferenceField component
 */
export interface ReferenceFieldProps<T> {
  fieldName: keyof T
  fieldKey?: string
  fieldSetting: ReferenceFieldSetting
  defaultValueIdOrPath?: string | number
  fetchItems: (fetchQuery: Query<T>) => Promise<T[]>
  onQueryChange: (key: string, query: Query<T>) => void
  renderSuggestion?: RenderSuggestion<T>
}

/**
 * State object for the ReferenceField component
 */
export interface ReferenceFieldState<T extends GenericContent> {
  inputValue: string
  isLoading: boolean
  isOpened: boolean
  term?: string
  items: T[]
  selected: T | null
  anchorEl: HTMLElement
  renderSuggestion: RenderSuggestion<T>
}

/**
 * Reference field picker component
 */
export class ReferenceField<T extends GenericContent = GenericContent> extends React.Component<
  ReferenceFieldProps<T> & MaterialTextFieldProps,
  ReferenceFieldState<T>
> {
  /**
   *
   */
  constructor(props: ReferenceField<T>['props']) {
    super(props)
    this.handleSelect = this.handleSelect.bind(this)
    this.fetchItems = debounce(this.fetchItems.bind(this), 500)
    this.onChange = this.onChange.bind(this)
  }

  /**
   * Initial state
   */
  public state: ReferenceFieldState<T> = {
    inputValue: '',
    isOpened: false,
    isLoading: false,
    items: [],
    selected: null,
    anchorEl: null as any,
    renderSuggestion: (item, { query, isHighlighted }) => {
      const matchesName = match(item.DisplayName || item.Name, query)
      const partsName = parse(item.DisplayName || item.Name, matchesName)

      const primary = partsName.map((part, index) =>
        part.highlight ? (
          <span key={String(index)} style={{ fontWeight: 500 }}>
            {part.text}
          </span>
        ) : (
          <strong key={String(index)} style={{ fontWeight: 300 }}>
            {part.text}
          </strong>
        ),
      )

      const matchesPath = match(item.Path, query)
      const partsPath = parse(item.Path, matchesPath)

      const secondary = partsPath.map((part, index) =>
        part.highlight ? (
          <span key={String(index)} style={{ fontWeight: 500 }}>
            {part.text}
          </span>
        ) : (
          <strong key={String(index)} style={{ fontWeight: 300 }}>
            {part.text}
          </strong>
        ),
      )

      return (
        <ListItem key={item.Id} selected={isHighlighted} component="div" value={item.Id}>
          <ListItemText primary={primary} secondary={secondary} />
          <Typography variant="body2">{}</Typography>
        </ListItem>
      )
    },
  }

  public onChange = (inputValue: string) => {
    this.setState({ inputValue })
    this.fetchItems(this.getQueryFromTerm(`*${inputValue}*`))
  }

  public async fetchItems(query: Query<T>) {
    const items = await this.props.fetchItems(query)
    this.setState({ items })
  }

  public renderInputComponent: RenderInputComponent<T> = (inputProps: InputProps<T>) => {
    const {
      classes,
      inputRef = () => {
        /** */
      },
      ref,
      defaultValue,
      onChange,
      displayName,
      name,
      ...other
    } = inputProps

    return (
      <TextField
        type="text"
        label={this.props.fieldSetting && this.props.fieldSetting.DisplayName}
        placeholder={this.props.fieldSetting && this.props.fieldSetting.DisplayName}
        title={this.props.fieldSetting && this.props.fieldSetting.Description}
        InputProps={{
          inputRef: node => {
            ref(node)
            inputRef(node)
          },
          endAdornment: this.state.isLoading ? (
            <InputAdornment position="end">
              <CircularProgress size={16} />
            </InputAdornment>
          ) : null,
        }}
        value={this.state.inputValue}
        onChange={ev => this.onChange(ev.target.value)}
        {...other}
      />
    )
  }
  public getSuggestionValue: GetSuggestionValue<T> = c => {
    return c.DisplayName || c.Name
  }
  public onSuggestionSelected: OnSuggestionSelected<T> = (_ev, data) => {
    _ev.preventDefault()
    this.handleSelect(data.suggestion)
  }

  public async componentDidMount() {
    /** */
    if (this.props.defaultValueIdOrPath) {
      const items = await this.props.fetchItems(
        new Query(q =>
          isNaN(this.props.defaultValueIdOrPath as number)
            ? q.equals('Path', this.props.defaultValueIdOrPath)
            : q.equals('Id', this.props.defaultValueIdOrPath),
        ),
      )
      if (items.length === 1 && items[0]) {
        const item = items[0]
        this.setState({
          inputValue: item.DisplayName || item.Name,
          selected: item,
        })
        this.props.onQueryChange(
          this.props.fieldKey || this.props.fieldName.toString(),
          new Query(q => q.equals(this.props.fieldName, item.Id)),
        )
      }
    }
  }

  public getQueryFromTerm<TQueryReturns>(term: string) {
    // tslint:disable
    const query = new Query<TQueryReturns>(q =>
      q.query(q2 =>
        q2
          .equals('Name', term)
          .or.equals('DisplayName', term)
          .or.equals('Path', term),
      ),
    )

    if (this.props.fieldSetting.AllowedTypes) {
      new QueryOperators(query).and.query(q2 => {
        ;(this.props.fieldSetting.AllowedTypes as string[]).map((allowedType, index, array) => {
          new QueryExpression(q2['queryRef']).term(`TypeIs:${allowedType}`)
          if (index < array.length - 1) {
            new QueryOperators(q2['queryRef']).or
          }
        })
        return q2
      })
    }

    if (this.props.fieldSetting.SelectionRoots && this.props.fieldSetting.SelectionRoots.length) {
      new QueryOperators(query).and.query(q2 => {
        ;(this.props.fieldSetting.SelectionRoots as string[]).forEach((root, index, array) => {
          new QueryExpression(q2['queryRef']).inTree(root)
          if (index < array.length - 1) {
            new QueryOperators(q2['queryRef']).or
          }
        })
        return q2
      })
    }
    return query
    // tslint:enable
  }

  private handleSelect(item: T) {
    this.setState({
      inputValue: item.DisplayName || item.Name,
      selected: item,
      isOpened: false,
    })
    this.props.onQueryChange(
      this.props.fieldKey || this.props.fieldName.toString(),
      new Query(q => q.equals(this.props.fieldName, item.Id)),
    )
  }

  /**
   * Renders the component
   */
  public render() {
    const displayName = (this.props.fieldSetting && this.props.fieldSetting.DisplayName) || this.props.label
    const description = (this.props.fieldSetting && this.props.fieldSetting.Description) || ''
    const inputProps: InputProps<T> = {
      value: this.state.inputValue,
      onChange: ev => this.onChange((ev.target as HTMLInputElement).value),
      id: 'CommandBoxInput',
      displayName,
      description,
    }

    return (
      <div ref={ref => ref && this.state.anchorEl !== ref && this.setState({ anchorEl: ref })}>
        <Autosuggest
          theme={{
            suggestionsList: {
              listStyle: 'none',
              margin: 0,
              padding: 0,
            },
          }}
          suggestions={this.state.items}
          onSuggestionsFetchRequested={async req => {
            const query = this.getQueryFromTerm<T>(`*${req.value}*`)
            const items = await this.props.fetchItems(query)
            this.setState({ items })
          }}
          onSuggestionsClearRequested={() => {
            this.setState({ items: [] })
          }}
          getSuggestionValue={this.getSuggestionValue}
          renderSuggestion={this.state.renderSuggestion}
          inputProps={inputProps}
          renderInputComponent={this.renderInputComponent}
          renderSuggestionsContainer={options => (
            <Paper
              square={true}
              style={{
                position: 'fixed',
                zIndex: 1,
              }}>
              <List component="nav" {...options.containerProps} style={{ padding: 0 }}>
                {options.children}
              </List>
            </Paper>
          )}
          onSuggestionSelected={this.onSuggestionSelected}
        />
      </div>
    )
  }
}
