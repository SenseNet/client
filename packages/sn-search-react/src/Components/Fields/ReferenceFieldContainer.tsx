import List from '@material-ui/core/List'
import Paper from '@material-ui/core/Paper'
import React from 'react'
import { RenderSuggestionsContainerParams } from 'react-autosuggest'

/**
 * Default container for Reference Field suggestion list
 * @param options
 */
export const ReferenceFieldContainer: React.FunctionComponent<RenderSuggestionsContainerParams> = options => (
  <Paper
    square={true}
    style={{
      position: 'fixed',
      zIndex: 1,
    }}>
    <List component="nav" {...options.containerProps} style={{ padding: 0 }}>
      {options.children}
    </List>
  </Paper>
)
