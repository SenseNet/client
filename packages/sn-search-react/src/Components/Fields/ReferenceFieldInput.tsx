import { TextField } from '@material-ui/core'
import React from 'react'
import { RenderInputComponentProps } from 'react-autosuggest'

type ReferenceFieldInputProps = {
  inputProps: RenderInputComponentProps
  displayName?: string
  helperText?: string
  description?: string
  autoFocus?: boolean
}
/**
 * Default Input field for Reference picker
 */
export const ReferenceFieldInput = (props: ReferenceFieldInputProps) => {
  const { description, displayName, helperText, inputProps } = props
  return (
    <TextField
      data-test={'reference-input'}
      autoFocus={props.autoFocus}
      type="text"
      label={displayName}
      placeholder={displayName}
      title={description}
      helperText={helperText}
      value={inputProps.value}
      InputProps={{
        ...(inputProps as any),
      }}
      onBlur={(ev) => inputProps.onBlur?.(ev)}
      onChange={(ev) => inputProps.onChange && inputProps.onChange(ev)}
    />
  )
}
