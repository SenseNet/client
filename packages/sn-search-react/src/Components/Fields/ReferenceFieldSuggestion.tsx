import ListItem from '@material-ui/core/ListItem'
import ListItemText from '@material-ui/core/ListItemText'
import { GenericContent } from '@sensenet/default-content-types'
import match from 'autosuggest-highlight/match'
import parse from 'autosuggest-highlight/parse'
import React from 'react'

/**
 * Parses a full text based on a search term and returns the corresponding list of JSX.Elements
 * @param text The full text to be parsed
 * @param term The search term
 */
export const getMatchParts = (text: string, term: string) => {
  const matchValue = match(term, text)
  const parseValue = parse(term, matchValue)

  return parseValue.map((part, index) =>
    part.highlight ? <strong key={String(index)}>{part.text}</strong> : <span key={String(index)}>{part.text}</span>,
  )
}
/**
 * Default Suggsetion component for Reference Field
 * @param props
 */
export const ReferenceFieldSuggestion: React.FunctionComponent<{
  item: GenericContent
  query: string
  isHighlighted: boolean
}> = props => {
  const primary = getMatchParts(props.query, props.item.DisplayName || props.item.Name)
  const secondary = getMatchParts(props.query, props.item.Path)
  return (
    <ListItem key={props.item.Id} selected={props.isHighlighted} value={props.item.Id}>
      <ListItemText primary={primary} secondary={secondary} />
    </ListItem>
  )
}
