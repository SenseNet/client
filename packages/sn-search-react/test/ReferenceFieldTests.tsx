import { TextField } from '@material-ui/core'
import { SchemaStore } from '@sensenet/client-core/dist/Schemas/SchemaStore'
import { GenericContent, ReferenceFieldSetting, SchemaStore as defaultSchemas } from '@sensenet/default-content-types'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import * as React from 'react'
import { ReferenceField } from '../src/Components/Fields/ReferenceField'

describe('ReferenceField Component', () => {
  configure({ adapter: new Adapter() })

  const schemaStore = new SchemaStore()
  schemaStore.setSchemas(defaultSchemas)
  const exampleSchema = schemaStore.getSchema(GenericContent)
  const exampleFieldSetting = exampleSchema.FieldSettings.find(f => f.Name === 'CreatedBy') as ReferenceFieldSetting

  it('Should be constructed', () => {
    shallow(
      <ReferenceField<GenericContent>
        fieldName="CreatedBy"
        fieldSetting={exampleFieldSetting}
        fetchItems={async () => []}
        onQueryChange={() => {
          /** */
        }}
      />,
    ).unmount()
  })

  it('Should be constructed with default Id', done => {
    shallow(
      <ReferenceField<GenericContent>
        fieldName="CreatedBy"
        fieldSetting={exampleFieldSetting}
        defaultValueIdOrPath={1}
        fetchItems={async fetchQuery => {
          expect(fetchQuery.toString()).toBe("Id:'1'")
          done()
          return [{ Id: 1, Name: 'a', Path: '', Type: 'Document' }]
        }}
        onQueryChange={() => {
          /** */
        }}
      />,
    )
  })

  it('Should be constructed with default Path', done => {
    shallow(
      <ReferenceField<GenericContent>
        fieldName="CreatedBy"
        fieldSetting={exampleFieldSetting}
        defaultValueIdOrPath="Root/Example/A"
        fetchItems={async fetchQuery => {
          expect(fetchQuery.toString()).toBe("Path:'Root/Example/A'")
          done()
          return [{ Id: 1, Name: 'a', Path: '', Type: 'Document' }]
        }}
        onQueryChange={() => {
          /** */
        }}
      />,
    )
  })

  it('Text change should trigger the fetchItems method', done => {
    const instance = shallow(
      <ReferenceField<GenericContent>
        fieldName="CreatedBy"
        fieldSetting={exampleFieldSetting}
        fetchItems={async fetchQuery => {
          expect(fetchQuery.toString()).toBe("(Name:'*a*' OR DisplayName:'*a*')")
          done()
          instance.unmount()
          return []
        }}
        onQueryChange={() => {
          /** */
        }}
      />,
    )
    instance
      .find(TextField)
      .props()
      .onChange({
        target: { value: 'a' },
        persist: () => {
          /** */
        },
      })
  })

  it('Text change query should include the allowed types', done => {
    const fieldSetting = { ...exampleFieldSetting }
    fieldSetting.AllowedTypes = ['User', 'Task']
    const instance = shallow(
      <ReferenceField<GenericContent>
        fieldName="CreatedBy"
        fieldSetting={fieldSetting}
        fetchItems={async fetchQuery => {
          expect(fetchQuery.toString()).toBe("(Name:'*a*' OR DisplayName:'*a*') AND (TypeIs:User OR TypeIs:Task)")
          done()
          instance.unmount()
          return []
        }}
        onQueryChange={() => {
          /** */
        }}
      />,
    )
    instance
      .find(TextField)
      .props()
      .onChange({
        target: { value: 'a' },
        persist: () => {
          /** */
        },
      })
  })

  it('Text change query should include the selection roots', done => {
    const fieldSetting = { ...exampleFieldSetting }
    fieldSetting.SelectionRoots = ['Root/A', 'Root/B']
    const instance = shallow(
      <ReferenceField<GenericContent>
        fieldName="CreatedBy"
        fieldSetting={fieldSetting}
        fetchItems={async fetchQuery => {
          expect(fetchQuery.toString()).toBe(
            '(Name:\'*a*\' OR DisplayName:\'*a*\') AND (InTree:"Root/A" OR InTree:"Root/B")',
          )
          done()
          instance.unmount()
          return []
        }}
        onQueryChange={() => {
          /** */
        }}
      />,
    )
    instance
      .find(TextField)
      .props()
      .onChange({
        target: { value: 'a' },
        persist: () => {
          /** */
        },
      })
  })
})
