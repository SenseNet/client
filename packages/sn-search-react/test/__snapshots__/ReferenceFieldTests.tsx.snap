// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReferenceField Component Container Should be rendered without error 1`] = `
<WithStyles(ForwardRef(Paper))
  square={true}
  style={
    Object {
      "position": "fixed",
      "zIndex": 1,
    }
  }
>
  <WithStyles(ForwardRef(List))
    className=""
    component="nav"
    id="1"
    key="1"
    role=""
    style={
      Object {
        "padding": 0,
      }
    }
  >
    <span>
      a
    </span>
  </WithStyles(ForwardRef(List))>
</WithStyles(ForwardRef(Paper))>
`;

exports[`ReferenceField Component InputField Should be rendered without error 1`] = `
<WithStyles(ForwardRef(TextField))
  InputProps={
    Object {
      "inputRef": [Function],
    }
  }
  onChange={[Function]}
  type="text"
  value=""
/>
`;

exports[`ReferenceField Component Suggestion Should be rendered without error 1`] = `
<WithStyles(ForwardRef(ListItem))
  button={true}
  key="1"
  selected={true}
>
  <WithStyles(ForwardRef(ListItemText))
    primary={
      Array [
        <strong>
          Al
        </strong>,
        <span>
          ma
        </span>,
      ]
    }
    secondary={
      Array [
        <span>
          Root/Sites/
        </span>,
        <strong>
          Al
        </strong>,
        <span>
          ma
        </span>,
      ]
    }
  />
</WithStyles(ForwardRef(ListItem))>
`;
